Websocket support

jukkar opened the pull request:
Initial websocket server side support.

jukkar committed:
net: websocket: Initial support for server websocket

This commit creates a websocket library that can be used by
applications. The websocket library implements currently only
server role and it uses services provided by net-app API.
The library supports TLS if enabled in configuration file.

This also adds websocket calls to HTTP app server if websocket
connection is established.

Signed-off-by: Jukka Rissanen <jukka.rissanen@linux.intel.com>

jukkar committed:
samples: net: websocket echo-server application

This is a http(s) server that supports also websocket.
It sends back any data sent to it over a websocket.

Signed-off-by: Jukka Rissanen <jukka.rissanen@linux.intel.com>

dbkinder commented on include/net/websocket.h:
```diff
@@ -0,0 +1,724 @@
+ * @param ctx Websocket context.
+ * @param pkt Network packet to send. If *pkt is NULL, then the API
+ * will allocate network packet and place data into it.
+ * @param type Type of the data (text or binary)
```
there is no parameter "type" or "mask"

dbkinder commented on samples/net/ws_echo_server/README.rst:
```diff
@@ -0,0 +1,36 @@
+    $ cd $ZEPHYR_BASE/samples/net/websocket
+    $ make pristine && make run
+
+Make can select the default configuration file based on the BOARD you've
```
I'd change this to:

      The default make BOARD configuration for this sample is ``qemu_x86``.

dbkinder reviewed (changes requested):
doxygen comment problem, and a suggested edit for README

dbkinder commented on include/net/websocket.h:
```diff
@@ -0,0 +1,734 @@
+ * @param opcode Operation code (text, binary, ping ,pong ,close)
+ * @param mask Mask the data, see RFC 6455 for details
+ * @param final Is this final message for this data send
+ * @param flag Can be used to send ping/pong or close the connection
```
no parameter "flag"

dbkinder commented on include/net/websocket.h:
```diff
@@ -0,0 +1,734 @@
+static inline int ws_send_data_to_client(struct ws_ctx *ctx,
+					 struct net_pkt *pkt,
+					 enum ws_opcode opcode,
+					 bool final,
```
missing @param for final

dbkinder reviewed (changes requested):
couple of doxygen errors...

mike-scott commented on subsys/net/lib/websocket/websocket.c:
```diff
@@ -0,0 +1,1529 @@
+			      enum ws_connection_type type,
+			      struct net_app_ctx *app_ctx)
+{
+#if defined(CONFIG_NET_DEBUG_WS) && (CONFIG_SYS_LOG_NET_LEVEL > 2)
```
This block seems to be duplicated quite a lot.  Can we #define PORT_STR at the top?   Also, PORT_STR seems to indicate it's a string.  Not a sizeof a string.

jukkar replied:
I will refactor this

mike-scott commented on subsys/net/ip/net_shell.c:
```diff
@@ -1353,13 +1397,26 @@ int net_shell_cmd_http(int argc, char *argv[])
 		if (strcmp(argv[arg], "monitor") == 0) {
 			printk("Activating HTTP monitor. Type \"net http\" "
 			       "to disable HTTP connection monitoring.\n");
+#if defined(CONFIG_NET_DEBUG_HTTP_CONN) && defined(CONFIG_HTTP_SERVER)
```
There's a placeholder function defined (#define http_server_cb(...) NULL) when CONFIG_NET_DEBUG_HTTP_CONN && CONFIG_HTTP_SERVER aren't defined, but we're checking for them here?  Should we remove the placeholder or remove the checks here? (and below)

jukkar replied:
Yes, will fix

mike-scott commented on subsys/net/ip/net_shell.c:
```diff
@@ -1329,21 +1330,64 @@ static void http_server_cb(struct http_server_ctx *entry,
+#define http_server_cb(...) NULL
+#endif
+
+static void ws_server_cb(struct ws_ctx *entry, void *user_data)
```
Seems like we can remove some of the #if defined checks below if we create a placeholder for this function similar to http_server_cb() above.

jukkar replied:
Indeed, I will change this.

mike-scott commented on samples/net/ws_echo_server/prj.conf:
```diff
@@ -0,0 +1,48 @@
+CONFIG_WS_SERVER=y
+CONFIG_WS_TLS=n
+
+CONFIG_MBEDTLS=y
```
Should the MBEDTLS settings be removed from this .conf?

jukkar replied:
This is needed for sha1 and base64 functions that are used even if we do not have TLS supported.

pfalcon replied:
Perhaps a comment line in this prj.conf would help then.

jukkar commented:
Thanks David, I have tried couple of times to setup doxygen so I could do documentation pre-checks myself but I have so far failed.

jukkar commented on subsys/net/ip/utils.c:
```diff
@@ -518,3 +518,230 @@ bool net_header_fits(struct net_pkt *pkt, u8_t *hdr, size_t hdr_size)
+}
+#endif /* CONFIG_NET_IPV6 || CONFIG_NET_IPV4 */
+
+bool net_ipaddr_parse(const char *str, size_t str_len, struct sockaddr *addr)
```
I will address this in PR #1198 as that contains these patches. I was forced to add unrelated patches to this one in order to compile the code.

jukkar commented on tests/net/utils/src/main.c:
```diff
@@ -743,10 +753,383 @@ void run_net_addr_tests(void)
+		{
+			.address = "192.0.2.3:80/foobar",
+			.len = sizeof("192.0.2.3:80/foobar") - 1,
+			.verdict = false
```
This will not pass as the port number that is tried to parse is "80/foobar" which is not an integer.

pfalcon commented on subsys/net/lib/websocket/Kconfig:
```diff
@@ -0,0 +1,60 @@
+# SPDX-License-Identifier: Apache-2.0
+#
+
+menuconfig WS
```
I'd say this is too cryptic. Despite the long sub-option names, I'd suggest to call this at least WEBSOCK

jukkar replied:
I think WEBSOCK is too long as we would need to change the API names too and then this WEBSOCK is way too long. If we really want to have more characters here, then perhaps WSS (websocket service) or similar could be used.

pfalcon replied:
My concern is that prj.conf remained readable. To that end, I'd say it might be possible to untie API prefixes from Kconfig values, and actually to have CONFIG_WEBSOCK, but still CONFIG_WS_HTTP_SERVER_NUM_URLS. Of course, that's just an opinion.

jukkar replied:
I changed the main config from CONFIG_WS to CONFIG_WEBSOCKET, this is indeed more readable now. I left the other variables as is as you suggested.

nashif replied:
WS is indeed too cryptic, is this WEBSERVICE, WEBSITE, WORKSHOP, WALLSTREET or maybe WORKSTATION?

Please make it clear by using WEBSOCKET, I do not think it is too long

pfalcon reviewed (commented):
I made some comments, emulating an end user who'd be using this API, thus looking at Kconfig, then at the header.

I'm not sure how useful they would be. As I commented previously, I find it hard to review large code drops, and this is another one.

So, instead just few generic comments:

1. A websocket support shouldn't require bundled HTTP server, instead it should integrate with an existing one.
2. If there's a desire for another HTTP server implementation, it should be a separate patch.
3. Description of this PR says "Eventually we might consider deprecating existing HTTP library and API and replace it by this library." But I fail to see any noticeable improvements over existing HTTP server. For example, this one still works in terms of network packets, which doesn't make sense for HTTP and far less for WebSockets. This surely leads to a bunch of hidden bugs, and few obvious and well-known by now ones, e.g. ws_add_http_data() is described as "Add HTTP data to the message", but that's not true, it either adds *very small amount of data*, or fails (silently or semi-silently), due to a well-know bug of not handling MTU fragmentation.

So, in my list, this patch neither works, nor can't be made to work without huge effort from different parties, which would require substantial redesign. But without complete redesign, it still will remain hard to use. Might work for toy examples as is, of course. Sorry if that sounds too critical, and I hope I'm wrong ;-). In the meantime, I'll better concentrate of IP level of things (unless explicitly assigned to work on application protocols).

jukkar commented:
> So, instead just few generic comments:

Hmm, I see lot of resistance here which I do not really understand.

> 
>     A websocket support shouldn't require bundled HTTP server, instead it should integrate with an existing one.

Unfortunately the current HTTP server is a bit cumbersome to use and integrating websocket there is not that easy. So I created a "new" HTTP server implementation that uses net-app API underneath, and then using websocket is much more easy. Also tried to address the issues we had with existing HTTP APIs, but it looks like I perhaps failed with this goal.

>     If there's a desire for another HTTP server implementation, it should be a separate patch.

Websocket is tightly integrated with HTTP, you cannot have WS without HTTP support.

>     Description of this PR says "Eventually we might consider deprecating existing HTTP library and API and replace it by this library." But I fail to see any noticeable improvements over existing HTTP server. For example, this one still works in terms of network packets, which doesn't make sense for HTTP and far less for WebSockets. This surely leads to a bunch of hidden bugs, and few obvious and well-known by now ones, e.g. ws_add_http_data() is described as "Add HTTP data to the message", but that's not true, it either adds very small amount of data, or fails (silently or semi-silently), due to a well-know bug of not handling MTU fragmentation.

Network packets are thing we should be using here. In HTTP APIs we tried to use buffers but I think that is wrong as then we need to convert to net_buf's anyway.

> 
> So, in my list, this patch neither works, nor can't be made to work without huge effort from different parties, which would require substantial redesign. But without complete redesign, it still will remain hard to use. Might work for toy examples as is, of course. Sorry if that sounds too critical, and I hope I'm wrong ;-). In the meantime, I'll better concentrate of IP level of things (unless explicitly assigned to work on application protocols).

I fail to see why this patch does not or will not work without huge effort. What parties are you talking about here? What kind of usage you are looking into?

Some background to the usecase for this websocket patch is the #980 RPL border router sample . Although we are not really in creating applications, the border router can be a nice demonstration tool for showing different network technologies that zephyr can provide. If one is only creating library APIs etc, it is easy to loose track how these APIs are used together, and for that we need applications.

jukkar commented on subsys/net/lib/websocket/Kconfig:
```diff
@@ -0,0 +1,60 @@
+	Enables websocket to be used over TLS link.
+
+config WS_HTTP_SERVER_NUM_URLS
+	int "Max number of HTTP URLs that the websocket server will handle"
```
The URL count is for both HTTP and WS. Even though there might be only one WS URL, there will definitely be more than one HTTP URL defined for the application. Please see examples for example in ws_echo_server or rpl_border_router samples how these are used.

jukkar commented on subsys/net/lib/websocket/Kconfig:
```diff
@@ -0,0 +1,60 @@
+	handle.
+
+config WS_HTTP_HEADER_FIELD_ITEMS
+	int "HTTP header field max number of items"
```
This is just copied from legacy HTTP server. We can change this of course.

jukkar commented on include/net/websocket.h:
```diff
@@ -0,0 +1,734 @@
+		struct ws_server_urls *urls;
+
+		/** HTTP URL parser */
+		struct http_parser_url parsed_uri;
```
You need to ask this from whoever sent the http parser patches to zephyr first time.

jukkar commented on include/net/websocket.h:
```diff
@@ -0,0 +1,734 @@
+		const char *url;
+
+		/** Length of the data in the request buf. */
+		size_t data_len;
```
Currently this supports only one request at a time, same issue with legacy HTTP API.

pfalcon replied:
This is one of many answers that proves there's not enough differences (improvements) over older HTTP API ;-).

jukkar replied:
There are actually. I have not yet sent it but I added multi connection support to net-app server yesterday.

jukkar commented on include/net/websocket.h:
```diff
@@ -0,0 +1,734 @@
+ * @param request_buf Caller-supplied buffer where the HTTP request will be
+ * stored
+ * @param request_buf_len Length of the caller-supplied buffer.
+ * @param server_banner Print information about started service. This is only
```
Just to be more user friendly, can be removed of course. We have this in legacy HTTP server too.

jukkar commented on include/net/websocket.h:
```diff
@@ -0,0 +1,734 @@
+ * @param ctx Websocket context.
+ * @param pkt Network packet to send. This can be left NULL if no user data
+ * is to be sent.
+ * @param opcode Operation code (text, binary, ping ,pong ,close)
```
Because it is a generic function that can send different kind of data (depending on opcode).

pfalcon replied:
I'd say, it sends a websock message ("ws_send_msg()"), which can be either data (text, binary) or command (everything else).

jukkar replied:
Sure

pfalcon commented:
> Hmm, I see lot of resistance here which I do not really understand.

I wouldn't call it "resistance", more "feeling of being lost", because I'm not sure where application protocol support in Zephyr goes.

> Websocket is tightly integrated with HTTP, you cannot have WS without HTTP support.

But you can use HTTP without WS! And this patch is structured in a way as to provide HTTP serving support as a special case of WebSocket server - APIs, config options are named like that. I can't agree that's right design. Instead, I'd write a generic, extensible HTTP server, which could delegate handling of Upgrade: requests to a WebSocket (or another) component.

> Network packets are thing we should be using here. In HTTP APIs we tried to use buffers but I think that is wrong as then we need to convert to net_buf's anyway.

Yes, but that's "conversion" is boring part of it, which should be hidden from user eyes. I actually grew to maintain even more orthodox point of view - for anything TCP related more complex than echo, working on the packet level is wrong.

> I fail to see why this patch does not or will not work without huge effort. What parties are you talking about here? What kind of usage you are looking into?

Parties which are interested in using HTTP server in Zephyr. I wouldn't say there were many, and some who tried, that I know, spotted various issues. Kind of usage? Well, can it serve 10KB page? Well, with https://github.com/zephyrproject-rtos/zephyr/pull/1330 it actually. Could. Ok, can it serve 1MB page? No, API is not designed for it. It supports only chunked encoding, it supports only one connection, etc., etc. - a lot of issues and limitations. Why can't we develop a generic HTTP server instead, then plug feature into it?

> Some background to the usecase for this websocket patch is the #980 RPL border router sample . Although we are not really in creating applications, the border router can be a nice demonstration tool for showing different network technologies that zephyr can provide. If one is only creating library APIs etc, it is easy to loose track how these APIs are used together, and for that we need applications.

Sure, just creating APIs isn't enough, but creating (single) adhoc samples may be not enough either - such one would be nice and cute, but someone trying to do something else and stepping away from what that samples does may hit "nothing works" situation. Why don't we develop a sample app which can serve 1MB page over plain old HTTP/1.0 for starters instead?


Anyway, that's just my opinion. I have nothing to show with those ideas (because HTTP server I tried to write hits IP stack issues, as I mentioned elsewhere), and you have this patch and even console over websocket. But in all fairness, I can't give +1 for this one, because I see that it's designed in too adhoc, not generic way (and it's 2nd such HTTP API in Zephyr). Won't say I'm "against" it either per above - my ideas are abstract, and you have real patches.

pfalcon commented:
Also, regarding "samples: net: common: Add rules to generate a header file" commit. My usual note that I wished that unrelated changes were submitted as separate PRs. (Yes, submitted on its own, it raises questions "Why do we need this?", which can be proactively answered by comment "Needed for PR #xxx"). Otherwise, remembering I saw it lately, I spent some time hunting thru different PRs to find it.


Anyway, lack of such functionality was one of the reasons I postponed submission of my socket http server sample. Well, now reading that commit, I'm not sure it describes how to use the added facility well enough. To show what I mean, suppose I have a binary file "foo.bar" and want to generate C syntax data for it as "kaboom.file", how does that facility help me? (It apparently doesn't, but then what it does and how to use it?)

Ideally, to make that work, I'd have:

~~~
obj-y += socket_dumb_http.o

socket_dumb_http.o: response.inc

response.inc: response.html
	bin2h 8 <$^ >$@
~~~

But that doesn't work. (And that's what annoying with Kbuild system - it's make-based, but doesn't work like make at all).

jukkar commented on include/net/websocket.h:
```diff
@@ -0,0 +1,734 @@
+};
+
+/**
+ * Websocket context information. This contains all the data that is
```
Currently this is server only. But it might be that we are able to combine the client one to this too.

jukkar commented:
> I wouldn't call it "resistance", more "feeling of being lost", because I'm not sure where application protocol support in Zephyr goes.

I do not know either. That is why I marked the ws support as experimental. But without patches, there is not much progress and we can talk on and on about it.

> 
>     Websocket is tightly integrated with HTTP, you cannot have WS without HTTP support.
> 
> But you can use HTTP without WS! And this patch is structured in a way as to provide HTTP serving support as a special case of WebSocket server - APIs, config options are named like that. I can't agree that's right design. Instead, I'd write a generic, extensible HTTP server, which could delegate handling of Upgrade: requests to a WebSocket (or another) component.

Indeed HTTP can be used without WS. And with little tweaking, this patchset supports that too. We could compile out the WS part from this if needed.

> 
>     Network packets are thing we should be using here. In HTTP APIs we tried to use buffers but I think that is wrong as then we need to convert to net_buf's anyway.
> 
> Yes, but that's "conversion" is boring part of it, which should be hidden from user eyes. I actually grew to maintain even more orthodox point of view - for anything TCP related more complex than echo, working on the packet level is wrong.
> 
>     I fail to see why this patch does not or will not work without huge effort. What parties are you talking about here? What kind of usage you are looking into?
> 
> Parties which are interested in using HTTP server in Zephyr. I wouldn't say there were many, and some who tried, that I know, spotted various issues. Kind of usage? Well, can it serve 10KB page? Well, with #1330 it actually. Could. Ok, can it serve 1MB page? No, API is not designed for it. It supports only chunked encoding, it supports only one connection, etc., etc. - a lot of issues and limitations. Why can't we develop a generic HTTP server instead, then plug feature into it?

Sure, just send patches.

> 
>     Some background to the usecase for this websocket patch is the #980 RPL border router sample . Although we are not really in creating applications, the border router can be a nice demonstration tool for showing different network technologies that zephyr can provide. If one is only creating library APIs etc, it is easy to loose track how these APIs are used together, and for that we need applications.
> 
> Sure, just creating APIs isn't enough, but creating (single) adhoc samples may be not enough either - such one would be nice and cute, but someone trying to do something else and stepping away from what that samples does may hit "nothing works" situation. Why don't we develop a sample app which can serve 1MB page over plain old HTTP/1.0 for starters instead?

Your example of sending 1mb over 1.0 connection is not a typical use case, and I am not sure if we really need to support that kind of feature.
For the current HTTP client and server APIs, I feel they are a failed experiment and require rewrite. There are lot of limitations there that should be fixed somehow. That is why I sent this WS patchset. It tries to fix some limitations, but with limited time I had, this is the current status. I decided not to touch existing HTTP APIs, but to do a rewrite using net_app API and see how far this can get us and possibly also fix issues that there are in net_app. One of the limitations in current net_app is the one connection limit, but I have a fix for that.

> 
> Anyway, that's just my opinion. I have nothing to show with those ideas (because HTTP server I tried to write hits IP stack issues, as I mentioned elsewhere), and you have this patch and even console over websocket. But in all fairness, I can't give +1 for this one, because I see that it's designed in too adhoc, not generic way (and it's 2nd such HTTP API in Zephyr). Won't say I'm "against" it either per above - my ideas are abstract, and you have real patches.

Nothing (probably other that time) prevents you from looking into this patchset, and improving it :)

jukkar commented:
> Also, regarding "samples: net: common: Add rules to generate a header file" commit. My usual note that I wished that unrelated changes were submitted as separate PRs. (Yes, submitted on its own, it raises questions "Why do we need this?", which can be proactively answered by comment "Needed for PR #xxx"). Otherwise, remembering I saw it lately, I spent some time hunting thru different PRs to find it.

I did not want to split the patchsets I have currently, into too small pieces. Working with these sets is very cumbersome in github, especially if the master is closed.

About how to use the this header file patchset:
1. Add this to application makefile (currently this is like this but perhaps this could be made more generic / automatic). This example is from border router sample app.

```
all: generate_inc_files generate_inc_gz_files
flash: generate_inc_files generate_inc_gz_files
run: generate_inc_files generate_inc_gz_files
pristine: remove_gen_files

BR_SRC = $(ZEPHYR_BASE)/samples/net/rpl_border_router/src
generate_inc_files:
	$(MAKE) -C $(BR_SRC) generate_inc_files Q=$(Q)
generate_inc_gz_files:
	$(MAKE) -C $(BR_SRC) generate_inc_gz_files Q=$(Q)
remove_gen_files:
	$(Q)rm -f src/file_*.h

```

2. Add needed binary/other embedded files into src/Makefile variables

```
# List of files that are used to generate .h file that can be included
# into .c file.
generate_inc_file += \
	echo-apps-cert.der \
	echo-apps-key.der

generate_inc_gz_file += \
	index.html \
	favicon.ico \
	br.js \
	style.css

```

3. In the application, do something with the embedded file

```
static int http_serve_index_html(struct ws_ctx *ctx)
{
#include "file_index.html_gz.h"

	NET_DBG("Sending index.html (%d bytes) to client",
		sizeof(file_index_html_gz));
	return http_response(ctx, HTTP_STATUS_200_OK_GZ, file_index_html_gz,
			     sizeof(file_index_html_gz));
}

```

The actual code that generates the include file can be found at samples/net/common/Makefile.common

jukkar commented on scripts/basic/stdin2hex.c:
```diff
@@ -0,0 +1,36 @@
+
+#include <stdio.h>
+
+int main(void)
```
You are looking the old code, this is already changed in latest version. Currently the latest version of this set is part of #980 and I will split the functionality from that PR after it matures a bit more. It is just PITA in github that I need to do these tweaks to get things reviewed in smaller and suitable chunks..

tbursztyka reviewed (approved):
no resistance from me ;)

jukkar commented:
> no resistance from me ;)

This websocket thing works but it might be better to separate the HTTP part from actual websocket implementation. I will send a new version that separates those two, so HTTP implementation would be done using net-app API, and then separate websocket API could be used together with the HTTP one.

jukkar commented:
Updated the websocket as HTTP library is using net-app API. This requires a bit more testing as I noticed there is some issue with TLS connection, but I will send it early in order to get feedback.

jukkar commented:
Fixed issues receiving long (>127) data. TLS also works ok. Tested with Firefox and wsdump.py tool.
This is ready to be reviewed so if possible, please review so that this could be added to 1.10 release.

jukkar commented:
Hi @pfalcon would you have some time to review this websocket support?

pfalcon commented on include/net/websocket.h:
```diff
@@ -0,0 +1,136 @@
+ *
+ * @details The function will remove websocket header from the network packet.
+ *
+ * @param pkt Network packet to send
```
Is "to send" expected here? I'd think the function is applied to recved packets.

pfalcon commented on include/net/websocket.h:
```diff
@@ -0,0 +1,136 @@
+ * @param pkt Network packet to process
+ * @param masking_value The mask value to use.
+ */
+void ws_mask_pkt(struct net_pkt *pkt, u32_t masking_value);
```
Looking just at the header, it's strange to have such a function as part of the public API. More specifically, I don't think it can work at all: masking_value is actually sequence on 4 bytes, applied with wrap-around to a stream of bytes. Thus, it needs to maintain an offset in 4-byte seq, or it won't be able to apply xform correctly to e.g. pkt of 5 bytes followed by pkt of 1 byte.

jukkar replied:
The function works just fine in my tests, it properly loops over masking_value. Why do you say it cannot work, have you tried it and it failed?

pfalcon replied:
No, it's based on the speculative review and knowledge of websocket protocol/experience with implementing it. There's always a chance of reviewer's error.

But what I see in https://github.com/zephyrproject-rtos/zephyr/pull/1271/commits/273e1130af5d38c0551f9906ab8ed1bcf0eb24e8#diff-4b79f0bbb3bc74c2d5bbeb4209f6490eR664 is a typical layering mis-assumption. Websocket records is a stream of bytes, wrapped in TCP stream of bytes, wrapped in IP packets. You receive an IP packet and think that it represents a complete Websocket message. It does not. 1st packet you receive contains 1st byte of Websocket header, 2nd byte - 2nd, etc., one packet per each byte. *That*'s the picture someone working with a streaming protocol like TCP should have in mind, not something else. So, you can't even "read" a mask value from a pkt - you should buffer a stream of bytes of enough length to hold an entire Websocket header, then when you fill it in, you should parse it. Then next packets won't even contain WS header, they will contain WS data (with arbitrary unmasking offset as was pointed out). Then finally one of the next headers will contain 1000 bytes of data followed by next WS header (in the same pkt).

(Yeah, when you implement the code infra to support correct handling of a stream protocol, you'll end up duplicating a lot of BSD Sockets functionality, that point was already shared.)

jukkar replied:
Yes, I see your point. I did not understand your first comment.

pfalcon commented on include/net/websocket.h:
```diff
@@ -0,0 +1,136 @@
+ * @return 0 if ok, 1 if there is no body, 2 if HTTP connection is to be
+ * upgraded to websocket one
+ */
+int ws_headers_complete(struct http_parser *parser);
```
Not sure if the function name corresponds to its functionality: "The function will check if this is a valid websocket connection.". If the function indeed does that, it should be named accordingly, otherwise, if name is right, does it have to be part of public API? (Maybe this header file needs to be split in "public API" and "private API" parts, if there's no choice of having 2 header files?)

jukkar replied:
Some of the functions could be marked as internal as they are very specific to the handshake phase. I will separate them.

pfalcon commented on subsys/net/lib/http/Kconfig:
```diff
@@ -39,9 +39,11 @@ config HTTP_HEADERS
 	int "HTTP header field max number of items"
 	depends on HTTP_SERVER
 	default 8
+	default 20 if WEBSOCKET
```
Is such a high value based on something? My visual memory says that no. of headers in websocket establishment request was under a dozen.

jukkar replied:
For example Firefox sends lot of headers, we need to be prepared to receive all of them as otherwise the handshake will fail.

pfalcon commented on subsys/net/lib/http/http_app_server.c:
```diff
@@ -647,16 +655,48 @@ static void http_received(struct net_app_ctx *app_ctx,
+
+		if (masked) {
+			/* Always deliver unmasked data to the application */
+			ws_mask_pkt(pkt, mask_value);
```
This won't work correctly, as mentioned in ws_mask_pkt() prototype review.

pfalcon commented on subsys/net/lib/websocket/websocket.c:
```diff
@@ -0,0 +1,485 @@
+{
+	int i;
+
+	for (i = 0; i < payload_len; i++) {
```
There should be `websocket_ctx` object, what is "i" in this function, should be part of this context.

jukkar replied:
There is no need for separate websocket_ctx, all context info is handled currently by http_ctx.

pfalcon replied:
Couple of points:

1. For a server implementation (e.g. HTTP server), there should be a context object. However, for each request it serves, there should be a separate *request context* object. The need for that should be obvious, considering a server may serve several independent requests (== clients). I had a feeling that https://github.com/zephyrproject-rtos/zephyr/pull/4243 doesn't have that, didn't raise it before to not procrastinate it further, but without a proper implementation of the above, it's an alpha version.

2. Websocket is a separate protocol from HTTP, so need a separate WS request (connection) context from HTTP request ctx. Implementation-wise, those can be the same underlying structure (i.e. if WS support enabled, every "mere HTTP" req ctx gets overhead of WS fields).

pfalcon commented on subsys/net/lib/websocket/websocket.c:
```diff
@@ -0,0 +1,485 @@
+void ws_mask_pkt(struct net_pkt *pkt, u32_t masking_value)
+{
+	struct net_buf *frag;
+	int i, count = 0;
```
Similar comment to the above (count belongs to websocket_ctx).

jukkar replied:
I lost the context here, what do you mean by this?

pfalcon reviewed (changes requested):
While I appreciate the idea of getting more useful protocols in the next release around the corner, per my review comments, I believe this doesn't implement a correct received data handling.

IMHO, instead of hasting and delivering half-bakes solutions, which will require a lot of rework going forward (possibly, breaking user APIs), we'd better do it bottom up, adding new layers on top of well-working existing layers. In this regard, I think the new HTTP API could use more work to get it more stable for the release, instead of hasting with e.g. Websocket support.

jukkar commented:
> While I appreciate the idea of getting more useful protocols in the next release around the corner, per my review comments, I believe this doesn't implement a correct received data handling.

I will send a new version shortly. And we can certainly postpone this to next version, I was not really expecting this to be accepted by you anyway. The API is currently marked as experimental so we should be able to change it at will. The client functionality would be nice to have and it might require changes to the APIs.

pfalcon commented:
> to be accepted by you anyway

I hope the keyword is "accepted" and not "by you" ;-). I really appreciate your work, but I don't think it scales to have a churn of one API implantation replace by another, repeated. So, I really think we should spend more homework on each piece (and in the process, learn the best way to layer things).

jukkar commented:
New version:
* moved some of the "internal" functions to separate header file that is not public
* changed the receiving side to continue reading stream until all data has been read

jukkar commented:
This needs still more testing, there was some issues when client is sending lot of data. I will continue with this tomorrow.

jukkar committed:
net: app: Do not allow local unspecified address in client

Do not allow :: or ANY address in client when sending data.

Signed-off-by: Jukka Rissanen <jukka.rissanen@linux.intel.com>

dbkinder commented on samples/net/ws_echo_server/README.rst:
```diff
@@ -0,0 +1,35 @@
+
+Build websocket-server sample application like this:
+
+.. code-block:: console
```
the cmake changes introduced a new way to document build steps

jukkar committed:
tests: net: websocket: Add tests for websocket

This tests websocket by creating a websocket support http server
and sending data to it and verifying the returned data is the same.

Signed-off-by: Jukka Rissanen <jukka.rissanen@linux.intel.com>

jukkar commented:
Updates:

* code to support one byte at a time handling
* readme file fix
* unit test added

Note that this is not to be applied to v1.10

jukkar commented:
Updates:

* Add IPv4 tests
* Fix sanity checks

nashif commented:
@jukkar  any update on this?

rveerama1 commented:
@nashif It's out of sync now. I am maintaining these patches in https://github.com/zephyrproject-rtos/zephyr/pull/5035/. But Websocket requires https://github.com/zephyrproject-rtos/zephyr/pull/6167 (http api change). So api change is not planned for 1.11, these PR's are now interdependent. @jukkar can you close this PR? we can create another one when https://github.com/zephyrproject-rtos/zephyr/pull/6167 is accepted to upstream.

jukkar commented:
Yes, lets close this one in order to avoid too much confusion.
